subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyCollector.java

```diff
@@ -93,7 +102,10 @@ private <D extends Dependency> void doAddLazy(Provider<D> dependency, @Nullable
             }
             return applyConfiguration((D) dep, config);
         });
-        getDependencies().add(provider);
+        getDependencies().addAll(providerFactory.provider(() -> {
+            Dependency value = provider.getOrNull();
+            return value != null ? Collections.singleton(value) : Collections.emptySet();
+        }));
```

Comment:
Why is this changing?

Reply:
I believe it is this issue. https://github.com/gradle/gradle/issues/20266

Can we pull this logic out into a generic `safeAddLater` method (for this usage and the one below) so its more clear this is a workaround for that issue?

Reply:
Pulled out and added to `DefaultSetProperty` so it's more visible for anyone who needs it and the eventual removal in the future.